spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/NDVSketchGenerator.java

```diff
@@ -0,0 +1,184 @@
+  }
+
+  private static StatisticsFile writeToPuffin(
+      Table table, List<String> columns, Map<String, ThetaSketchJavaSerializable> sketchMap)
+      throws IOException {
+    int columnSizes = columns.size();
+    TableOperations operations = ((HasTableOperations) table).operations();
+    FileIO fileIO = ((HasTableOperations) table).operations().io();
+    String path = operations.metadataFileLocation(String.format("%s.stats", UUID.randomUUID()));
+    OutputFile outputFile = fileIO.newOutputFile(path);
+    try (PuffinWriter writer =
+        Puffin.write(outputFile).createdBy("Spark DistinctCountProcedure").build()) {
+      for (int i = 0; i < columnSizes; i++) {
+        writer.add(
+            new Blob(
+                StandardBlobTypes.APACHE_DATASKETCHES_THETA_V1,
+                ImmutableList.of(table.schema().findField(columns.get(i)).fieldId()),
+                table.currentSnapshot().snapshotId(),
+                table.currentSnapshot().sequenceNumber(),
+                ByteBuffer.wrap(sketchMap.get(columns.get(i)).getSketch().toByteArray()),
+                null,
+                ImmutableMap.of()));
+      }
+      writer.finish();
+
+      return new GenericStatisticsFile(
+          table.currentSnapshot().snapshotId(),
+          path,
+          writer.fileSize(),
+          writer.footerSize(),
+          writer.writtenBlobsMetadata().stream()
+              .map(GenericBlobMetadata::from)
+              .collect(ImmutableList.toImmutableList()));
+
+    } catch (IOException e) {
+      throw e;
+    }
+  }
+
+  static Iterator<Tuple2<String, ThetaSketchJavaSerializable>> computeNDVSketches(
+      SparkSession spark, String tableName, long snapshotId, List<String> columns) {
+    String sql =
+        String.format(
+            "select %s from %s VERSION AS OF %d", String.join(",", columns), tableName, snapshotId);
+    Dataset<Row> data = spark.sql(sql);
+    final JavaPairRDD<String, String> pairs =
+        data.javaRDD()
+            .mapPartitionsToPair(
+                (PairFlatMapFunction<Iterator<Row>, String, String>)
+                    input -> {
+                      final List<Tuple2<String, String>> list = Lists.newArrayList();
+                      while (input.hasNext()) {
+                        final Row row = input.next();
+                        int size = row.size();
+                        for (int i = 0; i < size; i++) {
+                          list.add(new Tuple2<>(columns.get(i), row.get(i).toString()));
+                        }
+                      }
+                      return list.iterator();
+                    });
+
+    final JavaPairRDD<String, ThetaSketchJavaSerializable> sketches =
+        pairs.aggregateByKey(
+            new ThetaSketchJavaSerializable(),
+            1, // number of partitions
+            new Add(),
+            new Combine());
+
+    return sketches.toLocalIterator();
+  }
+
+  static class Add
```

Comment:
can we use lambdas here for cleaner code?  like
```
 (sketch, val) -> {
              sketch.update(val;);
              return sketch;
          },
```

The next one may be too complex to inline but maybe we can reduce the ugly java boilerplate